bundles/nl.asml.matala.product/src/nl/asml/matala/product/validation/ProductValidator.xtend

```diff
@@ -8,108 +8,263 @@ import nl.esi.comma.types.types.TypesPackage
 import org.eclipse.emf.common.util.URI
 import org.eclipse.xtext.EcoreUtil2
 import org.eclipse.xtext.validation.Check
+import nl.asml.matala.product.product.Function
+import java.util.HashSet
+import nl.asml.matala.product.product.ProductPackage
+import nl.asml.matala.product.product.Update
+import nl.asml.matala.product.product.Block
+import nl.esi.comma.expressions.expression.ExpressionVariable
+import nl.asml.matala.product.generator.ValidationHelper
+import java.util.Set
 
 /**
  * This class contains custom validation rules. 
- *
+ * 
  * See https://www.eclipse.org/Xtext/documentation/303_runtime_concepts.html#validation
  */
 class ProductValidator extends AbstractProductValidator {
-	@Check
+
+    @Check
     override checkImportForValidity(Import imp) {
-        if(!EcoreUtil2.isValidUri(imp, URI.createURI(imp.getImportURI()))) {
+        if (!EcoreUtil2.isValidUri(imp, URI.createURI(imp.getImportURI()))) {
             error("Invalid resource", imp, TypesPackage.eINSTANCE.getImport_ImportURI());
         } else {
             /*val Resource r = EcoreUtil2.getResource(imp.eResource, imp.importURI)
-            if(! (r.allContents.head instanceof InterfaceDefinition ||
-                r.allContents.head instanceof FeatureDefinition
-            ))
-                error("The imported resource is not an interface definition or a feature definition.", imp, TypesPackage.eINSTANCE.import_ImportURI)
-        }*/
+             * if(! (r.allContents.head instanceof InterfaceDefinition ||
+             *     r.allContents.head instanceof FeatureDefinition
+             * ))
+             *     error("The imported resource is not an interface definition or a feature definition.", imp, TypesPackage.eINSTANCE.import_ImportURI)
+             }*/
         }
     }
-    
-    
-    /* STRANGE BUG: Output Vars are Empty. Appears in Input Vars. 
-     * Not appearing as problem during product generation!
-    */
-    /*
+
+
+    /**
+     * Prevent Duplicate Updates in Actions
+     */
+     
     @Check
-    public void checkInputOutputMapping(Product p) 
-    {
-        var outVarMap = new HashMap<String,List<String>>();
-        var inVarMap  = new HashMap<String,List<String>>();
-        
-        for(var b : p.getBlock()) {
-            System.out.println(" + Parsing Block: " + b.getName());
-            for(var ov : b.getOutvars()) {
-                if(outVarMap.containsKey(b.getName())) {
-                    outVarMap.get(b.getName()).add(ov.getName());
-                    System.out.println("    Added Output: " + ov.getName());
-                }
-                else {
-                    var lst = new ArrayList<String>();
-                    lst.add(ov.getName());
-                    outVarMap.put(b.getName(), lst);
-                    System.out.println("    Added Output to Existing: " + ov.getName());
+    def preventDuplicationUpdatesActions(Function function) {
```

Comment:
The actual and only change of this pull request is that this method was added, right?

Did you move other methods within this file, leading to all these other modifications? If so, please be aware that this complicates reviewing your pull request, might complicate merging later on and therefore is not preferred